libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,613 @@
+ * The code has the start time of reionization as a free parameter: if this start of reionization is late then there
+ * will be a strong sudden burst of heating.
+ *
+ * (Loosely) based on lightup_QSOs.py from https://github.com/uptonsanderbeck/helium_reionization
+ * HeII_heating.py contains the details of the reionization history and how it is generated.
+ *
+ * This code should run only during a PM timestep, when all particles are active
+ * We lose time resolution but I cannot think of another way to ensure cooling is modelled correctly.
+ */
+
+/*
+ * 1. Use a BH accretion rate threshold rather than a BH mass threshold.
+ * 2. Bubble size correlates with black hole mass.
+ *
+ * 3. Add a switch to place bubbles at random in massive halos if there are no black holes.
+ * 4. Bubble size correlates with halo mass.
+ */
+
+#include <math.h>
+#include <mpi.h>
+#include <string.h>
+#include <gsl/gsl_interp.h>
+#include "cooling_qso_lightup.h"
+#include "physconst.h"
+#include "slotsmanager.h"
+#include "partmanager.h"
+#include "treewalk.h"
+#include "allvars.h"
+#include "hydra.h"
+#include "drift.h"
+#include "walltime.h"
+#include "fof.h"
+#include "utils/endrun.h"
+#include "utils/paramset.h"
+#include "utils/mymalloc.h"
+
+#define E0_HeII 54.4 /* HeII ionization potential in eV*/
+#define HEMASS 4.002602 /* Helium mass in amu*/
+
+/*Parameters for the quasar driven helium reionization model.*/
+struct qso_lightup_params
+{
+    int QSOLightupOn; /* Master flag enabling the helium reioization heating model.*/
+
+    double qso_candidate_min_mass; /* Minimum mass of a quasar halo candidate.
+                                  To become a quasar a FOF group should have a mass between min and max. */
+    double qso_candidate_max_mass; /* Minimum mass of a quasar halo candidate.*/
+
+    double mean_bubble; /* Mean size of the quasar bubble.*/
+    double var_bubble; /* Variance of the quasar bubble size.*/
+
+    double heIIIreion_finish_frac; /* When the desired ionization fraction exceeds this value,
+                                      the code will flash-ionize all remaining particles*/
+    double heIIIreion_start; /* Time at which start_reionization is called and helium III reionization begins*/
+};
+
+static struct qso_lightup_params QSOLightupParams;
+/* Memory for the helium reionization history. */
+/* Instantaneous heating from low-energy (short mean free path)
+ * photons to the Quasar to a newly ionized particle.
+ * Computed from parameters stored in the text file.
+ * In ergs.*/
+static double qso_inst_heating;
+static int Nreionhist;
+static double * He_zz;
+static double * XHeIII;
+static double * LMFP;
+static gsl_interp * HeIII_intp;
+static gsl_interp * LMFP_intp;
+/* Counter for particles ionized this step*/
+static int N_ionized;
+
+/*This is a helper for the tests*/
+void set_qso_lightup_par(struct qso_lightup_params qso)
```

Comment:
'set_qso_lightup_for_testing' -- if it really is tests-only.

Reply:
We have a convention in the other modules of naming functions like this in this way, so I'd like to keep it.